server/public/model/custom_profile_attributes.go

```diff
@@ -20,3 +27,159 @@ func CustomProfileAttributesPropertySortOrder(p *PropertyField) int {
+	CustomProfileAttributesPropertyAttrsVisibility = "visibility"
+)
+
+const (
+	CustomProfileAttributesValueTypeEmail = "email"
+	CustomProfileAttributesValueTypeURL   = "url"
+	CustomProfileAttributesValueTypePhone = "phone"
+)
+
+func IsKnownCustomProfilteAttributesValueType(valueType string) bool {
+	switch valueType {
+	case CustomProfileAttributesValueTypeEmail,
+		CustomProfileAttributesValueTypeURL,
+		CustomProfileAttributesValueTypePhone:
+		return true
+	}
+
+	return false
+}
+
+const (
+	CustomProfileAttributesVisibilityHidden  = "hidden"
+	CustomProfileAttributesVisibilityWhenSet = "when_set"
+	CustomProfileAttributesVisibilityAlways  = "always"
+
+	CustomProfileAttributesVisibilityDefault = CustomProfileAttributesVisibilityWhenSet
+)
+
+func IsKnownCustomProfilteAttributesVisibility(visibility string) bool {
+	switch visibility {
+	case CustomProfileAttributesVisibilityHidden,
+		CustomProfileAttributesVisibilityWhenSet,
+		CustomProfileAttributesVisibilityAlways:
+		return true
+	}
+
+	return false
+}
+
+type CustomProfileAttributesSelectOption struct {
+	ID    string `json:"id"`
+	Name  string `json:"name"`
+	Color string `json:"color"`
+}
+
+func (c CustomProfileAttributesSelectOption) GetID() string {
+	return c.ID
+}
+
+func (c CustomProfileAttributesSelectOption) GetName() string {
+	return c.Name
+}
+
+func (c *CustomProfileAttributesSelectOption) SetID(id string) {
+	c.ID = id
+}
+
+func (c CustomProfileAttributesSelectOption) IsValid() error {
+	if c.ID == "" {
+		return errors.New("id cannot be empty")
+	}
+
+	if !IsValidId(c.ID) {
+		return errors.New("id is not a valid ID")
+	}
+
+	if c.Name == "" {
+		return errors.New("name cannot be empty")
+	}
+
+	return nil
+}
+
+type CPAField struct {
+	PropertyField
+	Attrs CPAAttrs
+}
+
+type CPAAttrs struct {
+	Visibility string                                                `json:"visibility"`
+	SortOrder  string                                                `json:"sort_order"`
+	Options    PropertyOptions[*CustomProfileAttributesSelectOption] `json:"options"`
+	ValueType  string                                                `json:"value_type"`
+}
+
+func (c *CPAField) ToPropertyField() *PropertyField {
+	pf := c.PropertyField
+
+	pf.Attrs = StringInterface{
+		CustomProfileAttributesPropertyAttrsVisibility: c.Attrs.Visibility,
+		CustomProfileAttributesPropertyAttrsSortOrder:  c.Attrs.SortOrder,
+		CustomProfileAttributesPropertyAttrsValueType:  c.Attrs.ValueType,
+		PropertyFieldAttributeOptions:                  c.Attrs.Options,
+	}
+
+	return &pf
+}
+
+func (c *CPAField) Sanitize() *AppError {
+	switch c.Type {
+	case PropertyFieldTypeText:
+		if valueType := strings.TrimSpace(c.Attrs.ValueType); valueType != "" {
+			if !IsKnownCustomProfilteAttributesValueType(valueType) {
+				return NewAppError("ValidateCPAField", "app.custom_profile_attributes.unknown_value_type.app_error", map[string]any{"ValueType": valueType}, "", http.StatusUnprocessableEntity)
```

Comment:
Instead of using different error strings for each error type, let's move to a single error that accepts parameters like we did on [PropertyFields](https://github.com/mattermost/mattermost/blob/master/server/public/model/property_field.go#L64) for example

Reply:
We're sanitizing and validating the field in the same func. I'd suggest to either split the process in two or be explicit in the name with something like `SanitizeAndValidate` or similar